xdsl/interactive/app.py

```diff
@@ -40,17 +40,83 @@
 from xdsl.ir import MLContext
 from xdsl.parser import Parser
 from xdsl.passes import ModulePass, PipelinePass, get_pass_argument_names_and_types
+from xdsl.pattern_rewriter import PatternRewriter
 from xdsl.printer import Printer
 from xdsl.tools.command_line_tool import get_all_dialects, get_all_passes
+from xdsl.transforms import individual_rewrite
 from xdsl.transforms.mlir_opt import MLIROptPass
 from xdsl.utils.exceptions import PassPipelineParseError
 from xdsl.utils.parse_pipeline import PipelinePassSpec, parse_pipeline
 
 from ._pasteboard import pyclip_copy
 
+
+class IndividualRewrite(NamedTuple):
+    """
+    Type alias for a possible rewrite, described by an operation and pattern name.
+    """
+
+    operation: str
+    pattern: str
+
+
+class IndexedIndividualRewrite(NamedTuple):
+    """
+    Type alias for a specific rewrite pattern, additionally consisting of its operation index.
+    """
+
+    operation_index: int
+    rewrite: IndividualRewrite
+
+
 ALL_PASSES = tuple(sorted((p_name, p()) for (p_name, p) in get_all_passes().items()))
 """Contains the list of xDSL passes."""
 
+ALL_PATTERNS: tuple[IndividualRewrite, ...] = tuple(
+    IndividualRewrite(op_name, pattern_name)
+    for (op_name, pattern_by_name) in individual_rewrite.REWRITE_BY_NAMES.items()
+    for (pattern_name, _) in pattern_by_name.items()
+)
+"""Contains all the rewrite patterns."""
+
+
+def get_all_possible_rewrites(
+    patterns: tuple[IndividualRewrite, ...], op: ModuleOp
+) -> tuple[IndexedIndividualRewrite, ...]:
```

Comment:
can you please pass the REWRITE_BY_NAMES as an argument? that way the function doesn't depend on external state, and can be tested individually